README.md
# [lazy-arrayify][author-www-url] [![npmjs.com][npmjs-img]][npmjs-url] [![The MIT License][license-img]][license-url] [![npm downloads][downloads-img]][downloads-url]  

> We are lazy, also [lazy-cache][]d and [browserify][]-ready - just arrayify, falsey values returns empty array. In bonus with `.isArray` method.

[![code climate][codeclimate-img]][codeclimate-url] [![standard code style][standard-img]][standard-url] [![travis build status][travis-img]][travis-url] [![coverage status][coveralls-img]][coveralls-url] [![dependency status][david-img]][david-url]

## Install
```
npm i lazy-arrayify --save
```

## Usage
> For more use-cases see the [tests](./test.js)

```js
const lazyArrayify = require('lazy-arrayify')
```

### [.isArray](index.js#L59)
> Check if value is array using the [isarray][] module.

**Params**

* `val` **{Mixed}**    
* `returns` **{Boolean}**  

**Example**

```js
var isArray = require('lazy-arrayify').isArray

console.log(isArray(1234))        // => false
console.log(isArray('str'))       // => false
console.log(isArray(null))        // => false
console.log(isArray())            // => false
console.log(isArray(0))           // => false
console.log(isArray(false))       // => false
console.log(isArray([null, 123])) // => true
console.log(isArray([null]))      // => true
console.log(isArray([false]))     // => true
```

### [.arrayify](index.js#L85)
> Returns empty array on falsey values.

**Params**

* `val` **{Mixed}**    
* `returns` **{Array}**  

**Example**

```js
var arrayify = require('lazy-arrayify').arrayify

console.log(arrayify(1234))        // => [1234]
console.log(arrayify('str'))       // => ['str']
console.log(arrayify(null))        // => []
console.log(arrayify())            // => []
console.log(arrayify(0))           // => []
console.log(arrayify(false))       // => []
console.log(arrayify([null, 123])) // => [null, 123]
console.log(arrayify([false]))     // => [false]
```

## Related
* [arrify](https://www.npmjs.com/package/arrify): Convert a value to an array | [homepage](https://github.com/sindresorhus/arrify)
* [isarray](https://www.npmjs.com/package/isarray): Array#isArray for older browsers | [homepage](https://github.com/juliangruber/isarray)
* [lazy-utils](https://www.npmjs.com/package/lazy-utils): Most of the used mostly everywhere utils: [extend-shallow][], [isobject][], [is-extendable][] (used in… [more](https://www.npmjs.com/package/lazy-utils) | [homepage](https://github.com/tunnckocore/lazy-utils)
* [limon](https://www.npmjs.com/package/limon): The pluggable JavaScript lexer on per character basis. | [homepage](https://github.com/limonjs/limon)
* [limon-prev-next](https://www.npmjs.com/package/limon-prev-next): Plugin for [limon][] pluggable lexer that adds `prev` and `next` methods. | [homepage](https://github.com/limonjs/limon-prev-next)
* [postjson](https://www.npmjs.com/package/postjson): Transforming JSON with plugins. | [homepage](https://github.com/postjson/postjson)

## Contributing
Pull requests and stars are always welcome. For bugs and feature requests, [please create an issue](https://github.com/tunnckoCore/lazy-arrayify/issues/new).  
But before doing anything, please read the [CONTRIBUTING.md](./CONTRIBUTING.md) guidelines.

## [Charlike Make Reagent](http://j.mp/1stW47C) [![new message to charlike][new-message-img]][new-message-url] [![freenode #charlike][freenode-img]][freenode-url]

[![tunnckoCore.tk][author-www-img]][author-www-url] [![keybase tunnckoCore][keybase-img]][keybase-url] [![tunnckoCore npm][author-npm-img]][author-npm-url] [![tunnckoCore twitter][author-twitter-img]][author-twitter-url] [![tunnckoCore github][author-github-img]][author-github-url]

[lazy-cache]: https://github.com/jonschlinkert/lazy-cache
[browserify]: https://github.com/substack/node-browserify
[limon]: https://github.com/limonjs/limon
[isarray]: https://github.com/juliangruber/isarray
[extend-shallow]: https://github.com/jonschlinkert/extend-shallow
[isobject]: https://github.com/jonschlinkert/isobject
[is-extendable]: https://github.com/jonschlinkert/is-extendable

[npmjs-url]: https://www.npmjs.com/package/lazy-arrayify
[npmjs-img]: https://img.shields.io/npm/v/lazy-arrayify.svg?label=lazy-arrayify

[license-url]: https://github.com/tunnckoCore/lazy-arrayify/blob/master/LICENSE
[license-img]: https://img.shields.io/npm/l/lazy-arrayify.svg

[downloads-url]: https://www.npmjs.com/package/lazy-arrayify
[downloads-img]: https://img.shields.io/npm/dm/lazy-arrayify.svg

[codeclimate-url]: https://codeclimate.com/github/tunnckoCore/lazy-arrayify
[codeclimate-img]: https://img.shields.io/codeclimate/github/tunnckoCore/lazy-arrayify.svg

[travis-url]: https://travis-ci.org/tunnckoCore/lazy-arrayify
[travis-img]: https://img.shields.io/travis/tunnckoCore/lazy-arrayify/master.svg

[coveralls-url]: https://coveralls.io/r/tunnckoCore/lazy-arrayify
[coveralls-img]: https://img.shields.io/coveralls/tunnckoCore/lazy-arrayify.svg

[david-url]: https://david-dm.org/tunnckoCore/lazy-arrayify
[david-img]: https://img.shields.io/david/tunnckoCore/lazy-arrayify.svg

[standard-url]: https://github.com/feross/standard
[standard-img]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg

[author-www-url]: http://www.tunnckocore.tk
[author-www-img]: https://img.shields.io/badge/www-tunnckocore.tk-fe7d37.svg

[keybase-url]: https://keybase.io/tunnckocore
[keybase-img]: https://img.shields.io/badge/keybase-tunnckocore-8a7967.svg

[author-npm-url]: https://www.npmjs.com/~tunnckocore
[author-npm-img]: https://img.shields.io/badge/npm-~tunnckocore-cb3837.svg

[author-twitter-url]: https://twitter.com/tunnckoCore
[author-twitter-img]: https://img.shields.io/badge/twitter-@tunnckoCore-55acee.svg

[author-github-url]: https://github.com/tunnckoCore
[author-github-img]: https://img.shields.io/badge/github-@tunnckoCore-4183c4.svg

[freenode-url]: http://webchat.freenode.net/?channels=charlike
[freenode-img]: https://img.shields.io/badge/freenode-%23charlike-5654a4.svg

[new-message-url]: https://github.com/tunnckoCore/ama
[new-message-img]: https://img.shields.io/badge/ask%20me-anything-green.svg



CHANGELOG.md


## 1.0.3 - 2016-04-29
- Release v1.0.3 / npm@v1.0.3
- tweak some travis and other loading bugs

## 1.0.2 - 2016-04-29
- Release v1.0.2 / npm@v1.0.2
- add missing files from `files` field in pkg json

## 1.0.1 - 2016-04-29
- Release v1.0.1 / npm@v1.0.1
- allow accessing methods with `require('lazy-arrayify/isarray')` for example

## 1.0.0 - 2016-04-29
- Release v1.0.0 / npm@v1.0.0
- fix tests
- expose both `isArray` and `arrayify`
- docs
- implement

## 0.0.0 - 2016-04-29
- Initial commit

test.js
/*!
 * lazy-arrayify <https://github.com/tunnckoCore/lazy-arrayify>
 *
 * Copyright (c) 2016 Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

'use strict'

var test = require('assertit')
var lazy = require('./index')
var arrayify = require('./arrayify')
var isarray = require('./isarray')

test('should expose `.arrayify` and `.isArray` methods', function (done) {
  test.strictEqual(typeof lazy.arrayify, 'function')
  test.strictEqual(typeof lazy.isArray, 'function')
  test.strictEqual(typeof arrayify, 'function')
  test.strictEqual(typeof isarray, 'function')
  test.strictEqual(isarray(), false)
  test.strictEqual(isarray(0), false)
  test.strictEqual(isarray(null), false)
  test.strictEqual(isarray([]), true)
  test.strictEqual(isarray([null]), true)
  test.strictEqual(isarray([1, 2, 3]), true)
  done()
})

test('should return empty array on falsey values', function (done) {
  test.deepEqual(lazy.arrayify(null), [])
  test.deepEqual(lazy.arrayify(false), [])
  test.deepEqual(lazy.arrayify(undefined), [])
  test.deepEqual(lazy.arrayify(), [])
  test.deepEqual(lazy.arrayify(0), [])
  test.deepEqual(lazy.arrayify(''), [])
  test.deepEqual(arrayify(null), [])
  test.deepEqual(arrayify(false), [])
  test.deepEqual(arrayify(undefined), [])
  test.deepEqual(arrayify(), [])
  test.deepEqual(arrayify(0), [])
  test.deepEqual(arrayify(''), [])
  done()
})

test('should return array if array given', function (done) {
  test.deepEqual(lazy.arrayify(['']), [''])
  test.deepEqual(lazy.arrayify([1, 2, 3]), [1, 2, 3])
  test.deepEqual(lazy.arrayify([null, false]), [null, false])
  test.deepEqual(lazy.arrayify([0, undefined]), [0, undefined])
  test.deepEqual(arrayify(['']), [''])
  test.deepEqual(arrayify([1, 2, 3]), [1, 2, 3])
  test.deepEqual(arrayify([null, false]), [null, false])
  test.deepEqual(arrayify([0, undefined]), [0, undefined])
  done()
})

test('should arrayify truthy values', function (done) {
  test.deepEqual(lazy.arrayify(123), [123])
  test.deepEqual(lazy.arrayify(true), [true])
  test.deepEqual(lazy.arrayify('foo'), ['foo'])
  test.deepEqual(arrayify(123), [123])
  test.deepEqual(arrayify(true), [true])
  test.deepEqual(arrayify('foo'), ['foo'])
  done()
})


isarray.js
/*!
 * lazy-utils <https://github.com/tunnckoCore/lazy-utils>
 *
 * Copyright (c) 2016 Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

'use strict'

var utils = require('lazy-cache')(require)
var fn = require

require = utils // eslint-disable-line no-undef, no-native-reassign
require('isarray', 'isArray')
require = fn // eslint-disable-line no-undef, no-native-reassign

/**
 * > Check if value is array using the [isarray][] module.
 *
 * **Example**
 *
 * ```js
 * var isArray = require('lazy-arrayify').isArray
 *
 * console.log(isArray(1234))        // => false
 * console.log(isArray('str'))       // => false
 * console.log(isArray(null))        // => false
 * console.log(isArray())            // => false
 * console.log(isArray(0))           // => false
 * console.log(isArray(false))       // => false
 * console.log(isArray([null, 123])) // => true
 * console.log(isArray([null]))      // => true
 * console.log(isArray([false]))     // => true
 * ```
 *
 * @name   .isArray
 * @param  {Mixed} `val`
 * @return {Boolean}
 * @api public
 */

module.exports = function isArray (val) {
  return utils.isArray(val)
}


arrayify.js
/*!
 * lazy-utils <https://github.com/tunnckoCore/lazy-utils>
 *
 * Copyright (c) 2016 Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

'use strict'

var utils = require('lazy-cache')(require)
var fn = require

require = utils // eslint-disable-line no-undef, no-native-reassign
require('isarray', 'isArray')
require = fn // eslint-disable-line no-undef, no-native-reassign

/**
 * > Returns empty array on falsey values.
 *
 * **Example**
 *
 * ```js
 * var arrayify = require('lazy-arrayify').arrayify
 *
 * console.log(arrayify(1234))        // => [1234]
 * console.log(arrayify('str'))       // => ['str']
 * console.log(arrayify(null))        // => []
 * console.log(arrayify())            // => []
 * console.log(arrayify(0))           // => []
 * console.log(arrayify(false))       // => []
 * console.log(arrayify([null, 123])) // => [null, 123]
 * console.log(arrayify([false]))     // => [false]
 * ```
 *
 * @param  {Mixed} `val`
 * @return {Array}
 * @api public
 */

module.exports = function arrayify (val) {
  if (!val) return []
  if (!utils.isArray(val)) return [val]
  return val
}


index.js
/*!
 * lazy-utils <https://github.com/tunnckoCore/lazy-utils>
 *
 * Copyright (c) 2016 Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

'use strict'

/**
 * Module dependencies
 */

var utils = require('lazy-cache')(require)

/**
 * Temporarily re-assign `require` to trick browserify and
 * webpack into reconizing lazy dependencies.
 *
 * This tiny bit of ugliness has the huge dual advantage of
 * only loading modules that are actually called at some
 * point in the lifecycle of the application, whilst also
 * allowing browserify and webpack to find modules that
 * are depended on but never actually called.
 */

var fn = require
require = utils // eslint-disable-line no-undef, no-native-reassign

/**
 * Lazily required module dependencies
 */

/**
 * > Check if value is array using the [isarray][] module.
 *
 * **Example**
 *
 * ```js
 * var isArray = require('lazy-arrayify').isArray
 *
 * console.log(isArray(1234))        // => false
 * console.log(isArray('str'))       // => false
 * console.log(isArray(null))        // => false
 * console.log(isArray())            // => false
 * console.log(isArray(0))           // => false
 * console.log(isArray(false))       // => false
 * console.log(isArray([null, 123])) // => true
 * console.log(isArray([null]))      // => true
 * console.log(isArray([false]))     // => true
 * ```
 *
 * @name   .isArray
 * @param  {Mixed} `val`
 * @return {Boolean}
 * @api public
 */

require('isarray', 'isArray')

/**
 * > Returns empty array on falsey values.
 *
 * **Example**
 *
 * ```js
 * var arrayify = require('lazy-arrayify').arrayify
 *
 * console.log(arrayify(1234))        // => [1234]
 * console.log(arrayify('str'))       // => ['str']
 * console.log(arrayify(null))        // => []
 * console.log(arrayify())            // => []
 * console.log(arrayify(0))           // => []
 * console.log(arrayify(false))       // => []
 * console.log(arrayify([null, 123])) // => [null, 123]
 * console.log(arrayify([false]))     // => [false]
 * ```
 *
 * @name   .arrayify
 * @param  {Mixed} `val`
 * @return {Array}
 * @api public
 */

require('./arrayify', 'arrayify')

/**
 * Restore `require`
 */

require = fn // eslint-disable-line no-undef, no-native-reassign

/**
 * Expose `utils` modules
 */

module.exports = utils
